### ISSUE TITLE
Requesting Functorch Batching Rule for aten::linalg_eigh

### ISSUE THREAD

**Issue by Mark-Lowell** [pytorch/pytorch#108481]
### 🚀 The feature, motivation and pitch

I have a neural network that outputs (among other things) a 3-dimensional batch of matrices, representing the variance of a Gaussian distribution. I am enforcing that these matrices be symmetric positive-definite by applying a batched matrix exponent. As part of this calculation, I am performing a batch calculation of the eigenvalues and eigenvectors of the matrices, using functorch's excellent vmap operator. This triggers an internal warning asking me to file an issue to request a batching rule, so I'm doing that.

### Alternatives

Continue with the current implementation. That will work for me; this is not a performance bottleneck. I'm only submitting the issue because the warning asked me to.

### Additional context

Here is the warning message I received:

```
/content/utils.py:9: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::linalg_eigh. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at ../aten/src/ATen/functorch/BatchedFallback.cpp:82.)
  q, eigvals = la.eigh(x)
```

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @zou3519 @Chillee @samdow @kshitij12345 @janeyx99

**Comment by lezcano** [comment]
Two side notes:

`eigh` (and all the linalg functions) already support batches, so if you don't use `vmap` but regular batched tensors everything should work.

I would recommend against using `matrix_exp` to get a symmetric positive definite matrix, for the same reason I would recommend against using `exp` on a number to get a positive number: Numerical stability. `matrix_exp` is the same as computing the exponential of the eigenvalues, so you are probably going to face exploding gradient problems very quickly given that parametrisation. You may want into using a Cholesky-based decomposition, where you use something like `log(1+exp())` on the diagonal elements of lower triangular matrix and then compute `LL^T`. Thats the approach I implemented in [geotorch](https://github.com/Lezcano/geotorch/).

**Comment by Mark-Lowell** [comment]
> Two side notes:
> 
> `eigh` (and all the linalg functions) already support batches, so if you don't use `vmap` but regular batched tensors everything should work.

So it does! Thanks, I'd missed that.

> I would recommend against using `matrix_exp` to get a symmetric positive definite matrix, for the same reason I would recommend against using `exp` on a number to get a positive number: Numerical stability. `matrix_exp` is the same as computing the exponential of the eigenvalues, so you are probably going to face exploding gradient problems very quickly given that parametrisation. You may want into using a Cholesky-based decomposition, where you use something like `log(1+exp())` on the diagonal elements of lower triangular matrix and then compute `LL^T`. Thats the approach I implemented in [geotorch](https://github.com/Lezcano/geotorch/).

Thanks, I will take a look at that.
